containerd/nerdctl
# Volume QA part II: fixes and tests

This is part 2 of #3125 - same idea.

The original motivation was to significantly increase testing coverage & quality, convergence with docker behavior, and stability wrt concurrency for anything related to volume lifecycle.

In the process of doing so, several issues cropped-up, that this PR also fix.

Tests:
- parallelized (which does also serve as a very good way to uncover raciness)
- expanded
- cleaned-up / normalized
- added namespace testing
- test more thoroughly, both in terms of cases and validating content of stdout / stderr
- ~~move tests from _linux to all platforms (not sure why it was linux only in the first place)~~ - wish I had a windows env to test

Code changes:
- docker align: "no labels" should be `nil`, not `[]string{}`
- docker align: `--label ""` should error
- docker align: multi-inspect that is partly failing, and other partly failing conditions
- remove useless / duplicated code (specifically around validation of identifiers that was performed multiple times)
- make sure all methods of the volume store are locking (to avoid racy behavior)
- replace broken/racy logic "`get` then `create`" by just `create` (since behavior has been recently changed and now always returns success)
- make the underlying filesystem implementation private (eg: no longer expose `Path`) to prevent leaking implementation to the consumer (see for example `cmd/nerdctl/namespace.go b/cmd/nerdctl/namespace.go` that was reading the filesystem directly instead of calling `List`)
- added a "persistent" `Lock` and `Unlock` methods on the store, for operations that need to call multiple methods successively or need guarantees that volumes cannot be modified at all until they are done
- cleaned-up and expanded `lockutil`, so that we don't expose anymore `Flock` (unimplemented for non Unix), and instead expose new lock/unlock methods
- reviewed all the codebase for any cases where that global lock call is necessary and added them
- reviewed codebase to fix #3132 (and fix #3123 hopefully)

While this is a significant PR in term of apparent code change size, the added testing should give us a good confidence boost on what's happening wrt volumes.
